lib/block.ml

```diff
@@ -295,105 +296,195 @@ let version_enabled version =
 
 let get_label f (labels : Label.t list) = Util.List.find_map f labels
 
-let check_not_set msg = function
-  | Some _ -> Util.Result.errorf msg
-  | None -> Ok ()
+let label_not_allowed ~label ~kind =
+  Util.Result.errorf "`%s` label is not allowed for %s blocks." label kind
+
+let label_required ~label ~kind =
+  Util.Result.errorf "`%s` label is required for %s blocks." label kind
+
+let fmt_header = Fmt.using (Format.asprintf "`%a`" Header.pp) Fmt.string
+
+let header_not_allowed ~header ~kind =
+  Util.Result.errorf "header %a is not allowed for %s blocks."
+    (fun () -> Format.asprintf "%a" fmt_header)
+    header kind
+
+let header_required ~headers ~kind =
+  Util.Result.errorf "header %a is required for %s blocks."
+    (fun () ->
+      Format.asprintf "%a" (Fmt.list ~sep:(Fmt.unit " or ") fmt_header))
+    headers kind
 
 let check_no_errors = function
   | [] -> Ok ()
   | _ :: _ ->
       Util.Result.errorf "error block cannot be attached to a non-OCaml block"
 
-let mk ~line ~file ~column ~section ~labels ~legacy_labels ~header ~contents
-    ~errors =
-  let non_det =
-    get_label
-      (function
-        | Non_det (Some x) -> Some x
-        | Non_det None -> Some Label.default_non_det
-        | _ -> None)
-      labels
-  in
-  let part = get_label (function Part x -> Some x | _ -> None) labels in
-  let env = get_label (function Env x -> Some x | _ -> None) labels in
-  let dir = get_label (function Dir x -> Some x | _ -> None) labels in
-  let skip = List.exists (function Label.Skip -> true | _ -> false) labels in
-  let version =
-    get_label (function Version (x, y) -> Some (x, y) | _ -> None) labels
-  in
-  let source_trees =
-    List.filter_map
-      (function Label.Source_tree x -> Some x | _ -> None)
-      labels
-  in
-  let required_packages =
-    List.filter_map
-      (function Label.Require_package x -> Some x | _ -> None)
-      labels
-  in
-  let set_variables =
-    List.filter_map
-      (function Label.Set (v, x) -> Some (v, x) | _ -> None)
-      labels
-  in
-  let unset_variables =
-    List.filter_map (function Label.Unset x -> Some x | _ -> None) labels
-  in
-  let file_inc = get_label (function File x -> Some x | _ -> None) labels in
-  let open Util.Result.Infix in
-  ( match file_inc with
-  | Some file_included -> (
-      check_not_set
-        "`non-deterministic` label cannot be used with a `file` label." non_det
-      >>= fun () ->
-      check_not_set "`env` label cannot be used with a `file` label." env
-      >>= fun () ->
-      check_no_errors errors >>= fun () ->
+type label_config = {
```

Comment:
I think using a record here makes much more sense than what we use to have but `label_config` is a bit confusing. How about a simpler `labels` or eventually `block_config`?